### PR TITLE
Fix usage of EAV cache type

### LINKED ISSUE #32322: [Issue] Fix usage of EAV cache type
This issue is automatically created based on existing pull request: magento/magento2#31264: Fix usage of EAV cache type

---------

### Description (*)

This PR fixes usage of EAV cache type in some places

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. _Different frontend type can be set to EAV cache type without defining additional eav frontend type introduced by `configured_eav_cache` virtual type_
UPD: looks like it's bigger problem I see many other classes do not use `\Magento\Eav\Model\Cache\Type` and they can't be fixed due to `Magento\Framework\App\CacheInterface` and `Magento\Framework\Cache\FrontendInterface` incompatibilities.
It should be figured out and fixed in different PR/issue.
2. Class `\Magento\Catalog\Plugin\Model\ResourceModel\Config` checks if  EAV cache enabled but store data in default cache type. This PR fixes missed configuration

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by ilnytskyi**
### Description (*)

This PR fixes usage of EAV cache type in some places

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. _Different frontend type can be set to EAV cache type without defining additional eav frontend type introduced by `configured_eav_cache` virtual type_
UPD: looks like it's bigger problem I see many other classes do not use `\Magento\Eav\Model\Cache\Type` and they can't be fixed due to `Magento\Framework\App\CacheInterface` and `Magento\Framework\Cache\FrontendInterface` incompatibilities.
It should be figured out and fixed in different PR/issue.
2. Class `\Magento\Catalog\Plugin\Model\ResourceModel\Config` checks if  EAV cache enabled but store data in default cache type. This PR fixes missed configuration

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#32322: Fix usage of EAV cache type

### CODE REVIEW

**Comment by ilnytskyi** [comment]
@magento run all tests

File: app/code/Magento/Eav/Model/Cache/Type.php [hunk 1]
```diff
@@ -5,6 +5,7 @@
  */
 namespace Magento\Eav\Model\Cache;
 
+
```
Comment (lbajsarowicz) [inline app/code/Magento/Eav/Model/Cache/Type.php]: I don't think we need this extra line.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
🟡  I don't see any critical issues there.
Let's wait for Static Tests to finish, but the only issue is code style to adjust.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by ilnytskyi** [comment]
I think that in 
\Magento\Catalog\Plugin\Model\ResourceModel\Config
Wrong cache interface used or something
will try to figure out how to fix it.

**Comment by ilnytskyi** [comment]
@magento run all tests

**Comment by ilnytskyi** [comment]
I decided to open PR after I saw this change. But now i see it's a bit bigger problem.

https://github.com/magento/magento2/commit/4b66b00c574a0e092597f1603ede7bfcb8a95ed2#diff-b3fae67f1de2de925eb945c0dac8213f1f6f817053039c4d15eb698a338a3ac3

@vzabaznov could you take a look at this issue and help us figure out what to do.

I our project we use different cache backends for different cache types. Mainly to reduce data transfer from redis, and isolate containers in ASG and with 2.4.1 magento we faced the issue, that in RemoteSynchronized cache not all EAV keys were stored in containers. The first issue was the necessity to use a separate frontend with the name `eav`. And the other issue that, some classes still did not use proper cache types to store the date, e.g. `Magento\Catalog\Plugin\Model\ResourceModel\Config`

Probably the issue was not noticed since devs don't usually use this approach to point some caches to different storages, or do not debug if the keys actually come from proper storage.


This is serious issue because these docs pages do not really represent what actually going inside the app.
https://devdocs.magento.com/guides/v2.4/extension-dev-guide/cache/partial-caching/database-caching.html and 
https://devdocs.magento.com/guides/v2.4/extension-dev-guide/cache/partial-caching/database-caching.html#mage-cache-db-env

**Comment by ilnytskyi** [comment]
@maghamed could you explain some differences between 
`Magento\Framework\Cache\FrontendInterface` and `Magento\Framework\App\CacheInterface` and why `configured_eav_cache` virtual type used for EAV cache instead of `Magento\Eav\Model\Cache\Type` ?

As a dev I do not see point to define EAV cache type in XML https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Eav/etc/cache.xml and then use different classes that implement different interfaces to actually work with the cache storage.

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests

**Comment by engcom-Hotel** [comment]
@magento run WebAPI Tests

**Comment by lbajsarowicz** [comment]
Sorry, @engcom-Hotel - I'm no longer Maintainer.
@ihor-sviziev might be able to help

**Comment by ihor-sviziev** [comment]
It seems like this is a really complex issue in the area where I don't have enough expertise. Removing my assignee to allow bringing this PR by someone else

**Comment by gabrieldagama** [comment]
@magento create issue

**Comment by ilnytskyi** [comment]
up

**Comment by ilnytskyi** [comment]
up

**Comment by ilnytskyi** [comment]
up

**Comment by ihor-sviziev** [comment]
@ilnytskyi, could you please add more information on the changes? Does it improve performance? If so - could you share some improvement measurements you did?

**Comment by ilnytskyi** [comment]
@ihor-sviziev As I described in https://github.com/magento/magento2/pull/31264#issuecomment-743725776
It's mostly visible during using L2 cache. For EAV cache is written into target local storage, but  is fully loaded from default.
So instead loading a hash from shared storage and loading data from local storage, the cache always loads data from shared storage. It increases network usage.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Review by lbajsarowicz — APPROVED** [review]

**Comment by engcom-Dash** [comment]
Hello @ilnytskyi,

Thank you for your contributions!

Could you please provide the manual testing steps so we can proceed with this PR?

Thanks!

**Comment by ilnytskyi** [comment]
@engcom-Dash 

```
Class \Magento\Catalog\Plugin\Model\ResourceModel\Config checks if EAV cache enabled but store data in default cache type. This PR fixes missed configuration
```

basically the class checks if EAV cache enabled 
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Plugin/Model/ResourceModel/Config.php#L52

but actual read write happend on default cache type, not eav
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Plugin/Model/ResourceModel/Config.php#L73

can be checked via actual type of passed `CacheInterface $cache,`

**Comment by engcom-Dash** [comment]
@ilnytskyi - I have configured the L2 cache and checked the type of CacheInterface $cache, but I didn't notice any changes after implementing PR changes. Could you please let me know if I’m missing anything? Thanks.

Please see the Xdebug response for CacheInterface $cache below:
image

**Comment by ilnytskyi** [comment]
@engcom-Dash If you configure EAV to be stored in different storage than default.

Let's say default is Redis, and EAV is filesystem.

Maybe follow instruction from this docs - https://experienceleague.adobe.com/en/docs/commerce-operations/configuration-guide/cache/redis/redis-pg-cache#results 
As one can see different backends can be configured per cache type.

Default cache is always 
```
'backend' => '\\Magento\\Framework\\Cache\\Backend\\Redis',
```

and EAV

```
'backend' => 'Cm_Cache_Backend_File'
```

Are data actually writtend to the Redis or Filesystem for those keys?
Expected destination would be EAV to filesystem not to redis.

We found this issue when trying avoid using **L2** but just store EAV type in **tmpfs** on the server. But we noticed that EAV in those kays still used default destination.

**Comment by engcom-Dash** [comment]
@magento run all tests

File: app/code/Magento/Catalog/etc/di.xml [hunk 2]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0"?>
 <!--
 /**
- * Copyright © Magento, Inc. All rights reserved.
- * See COPYING.txt for license details.
+ * Copyright 2013 Adobe
+ * All Rights Reserved.
```
Comment (fballiano) [inline app/code/Magento/Catalog/etc/di.xml]: why this change? copyrighted code should not be merged in this repo

Comment (hostep) [reply app/code/Magento/Catalog/etc/di.xml]: @fballiano, see the thread in https://github.com/magento/magento2/pull/39099#pullrequestreview-2399484952 (especially my last comment there) and this comment here: https://github.com/magento/magento2/pull/37631#issuecomment-2430072090 and https://github.com/magento/magento-coding-standard/compare/v34...v35

So basically, it's a new requirement from magento's coding standards, an updated copyright header is required in every file touched from now on.

Comment (fballiano) [reply app/code/Magento/Catalog/etc/di.xml]: thanks @hostep, question: why does it say "all rights reserved", that doesn't sound open source or it sounds that that file is not under the OSS license

Comment (hostep) [reply app/code/Magento/Catalog/etc/di.xml]: No idea 🤷, you can take it up with @sidolov, he's on [Magento's Engineering Slack workspace](https://developer.adobe.com/open/magento/slack) if you want to discuss it in more detail.

**Comment by engcom-Dash** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Dash** [comment]
The Functional B2B test failures are different in recent 2 successive builds. They neither part of PR nor failing because of the PR changes, hence moving it to Merge in Progress.
Run 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/31264/541c1d300e91dd4344648d7eab80df0b/Functional/allure-report-b2b/index.html#categories/16b25a50b4a8b68f5548cdba82ebd5f8/43cc2cc92e4a0723/
image

Run 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/31264/61d013f3b977f3c8f51d8ebb9bc04fcc/Functional/allure-report-b2b/index.html#categories/a3bd7bb7965715d4b4c20d3e9cf75c2a/dd6bdc519ebcf74d/
image

**Pull request merged**